Add a channel to exit the remotecommand executor stream

ccchieh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
   https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
   -->

#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

Add a channel to exit the remotecommand executor stream.

By adding `CloseChan chan bool` to the StreamOptions struct, users can actively close the stream to prevent blocking.

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/client-go/issues/884

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:

- [KEP]: <link>

- [Usage]: <link>

- [Other doc]: <link>

  -->

ccchieh commented:
I signed it

ccchieh commented:
/assign @sttts

ccchieh commented:
triage/accepted

ccchieh commented:
/triage accepted

wzshiming commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -122,6 +123,11 @@ func (e *streamExecutor) Stream(options StreamOptions) error {
 	}
 	defer conn.Close()
 
+	go func() {
```
I think this PR change can't achieve the effect you want, but an extra goroutine leak

ccchieh replied:
> goroutine leak If CloseChan is not set

I found that the `Connection` interface provides a `CloseChan()`. Using `CloseChan()` can fix the problem of goroutine leakage.

```go
// Connection represents an upgraded HTTP connection.
type Connection interface {
	// CreateStream creates a new Stream with the supplied headers.
	CreateStream(headers http.Header) (Stream, error)
	// Close resets all streams and closes the connection.
	Close() error
	// CloseChan returns a channel that is closed when the underlying connection is closed.
	CloseChan() <-chan bool
	// SetIdleTimeout sets the amount of time the connection may remain idle before
	// it is automatically closed.
	SetIdleTimeout(timeout time.Duration)
}
```

wzshiming reviewed (changes requested):
The last two of the three commits you submitted are empty.

wzshiming commented:
``` diff
	req, err := http.NewRequest(e.method, e.url.String(), nil)
	if err != nil {
		return fmt.Errorf("error creating request: %v", err)
	}
+++	req = req.WithContext(ctx)

	conn, protocol, err := spdy.Negotiate(
		e.upgrader,
		&http.Client{Transport: e.transport},
		req,
		e.protocols...,
	)
```

Maybe this is better. WDYT

ccchieh reviewed (commented):
> ```diff
> 	req, err := http.NewRequest(e.method, e.url.String(), nil)
> 	if err != nil {
> 		return fmt.Errorf("error creating request: %v", err)
> 	}
> +++	req = req.WithContext(ctx)
> 
> 	conn, protocol, err := spdy.Negotiate(
> 		e.upgrader,
> 		&http.Client{Transport: e.transport},
> 		req,
> 		e.protocols...,
> 	)
> ```
> 
> Maybe this is better. WDYT

The context is really good, but adding a context will change the function parameters, and there may be incompatibility with the old code

ccchieh commented:
> The last two of the three commits you submitted are empty. please clean them up and force push.

@wzshiming The new code I added solved the leak problem and cleaned up the commit. Please check it out if you have time.

`Connection interface` provides CloseChan() 

```go
// Connection represents an upgraded HTTP connection.
type Connection interface {
	// CreateStream creates a new Stream with the supplied headers.
	CreateStream(headers http.Header) (Stream, error)
	// Close resets all streams and closes the connection.
	Close() error
	// CloseChan returns a channel that is closed when the underlying connection is closed.
	CloseChan() <-chan bool
	// SetIdleTimeout sets the amount of time the connection may remain idle before
	// it is automatically closed.
	SetIdleTimeout(timeout time.Duration)
	// RemoveStreams can be used to remove a set of streams from the Connection.
	RemoveStreams(streams ...Stream)
}
```

So when conn.CloseChan() receives the signal, it can exit.

fedebongio commented:
/triage accepted

wzshiming commented:
I think it might be more appropriate to add a method to the interface 

``` diff
diff --git a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
index cb39faf7f1a..ab1578de8b8 100644
--- a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
@@ -17,6 +17,7 @@ limitations under the License.
 package remotecommand
 
 import (
+	"context"
 	"fmt"
 	"io"
 	"net/http"
@@ -48,6 +49,8 @@ type Executor interface {
 	// is set, the stderr stream is not used (raw TTY manages stdout and stderr over the
 	// stdout stream).
 	Stream(options StreamOptions) error
+
+	StreamContext(ctx context.Context, options StreamOptions) error
 }
 
 type streamCreator interface {
@@ -106,7 +109,11 @@ func NewSPDYExecutorForProtocols(transport http.RoundTripper, upgrader spdy.Upgr
 // Stream opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects.
 func (e *streamExecutor) Stream(options StreamOptions) error {
-	req, err := http.NewRequest(e.method, e.url.String(), nil)
+	return e.StreamContext(context.Background(), options)
+}
+
+func (e *streamExecutor) StreamContext(ctx context.Context, options StreamOptions) error {
+	req, err := http.NewRequestWithContext(ctx, e.method, e.url.String(), nil)
 	if err != nil {
 		return fmt.Errorf("error creating request: %v", err)
 	}
```

ccchieh commented:
@wzshiming Your idea is better, I have updated the PR, thank you for your reminder.

wzshiming commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -48,6 +49,10 @@ type Executor interface {
 	Stream(options StreamOptions) error
+	// StreamContext initiates the transport of the standard shell streams through Context.
+	StreamContext(ctx context.Context, options StreamOptions) error
+	// StreamRequest initiates the transport of the standard shell streams through Request.
+	StreamRequest(req *http.Request, options StreamOptions) error
```
I don't think this method is necessary.

ccchieh replied:
I have changed it, thanks

wzshiming reviewed (commented):
Overall looks good to me.

PING @yliaog

wzshiming commented:
/ok-to-test

ccchieh commented:
/retest

wzshiming commented:
/lgtm

ccchieh commented:
@soltysh @yliaog Please help me check if there is any problem with this PR. Thank you.

yliaog commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -48,6 +49,8 @@ type Executor interface {
 	// stdout stream).
 	Stream(options StreamOptions) error
+	// StreamContext initiates the transport of the standard shell streams through Context.
+	StreamContext(ctx context.Context, options StreamOptions) error
```
this is breaking change, please send an email to kubernetes-sig-api-machinery@googlegroups.com to check if it's acceptable.

ccchieh replied:
Thanks for the tip, I have sent the email

lavalamp replied:
can you do a search to see if there are other implementations of this? If not it is probably fine to make this change.

pangsq commented:
Does the `StreamContext`  really work?

I tried a test and the context.CancelFunc did not stop the stream. If there is any wrong in the test code, please point it. Thanks

```golang
func attach2ServerContext(rawURL string, options StreamOptions) error {
	uri, _ := url.Parse(rawURL)
	exec, err := NewSPDYExecutor(&rest.Config{Host: uri.Host}, "POST", uri)
	if err != nil {
		return err
	}

	ctx, cancel := context.WithCancel(context.Background())
	e := make(chan error)
	go func(ctx context.Context, e chan error) {
		e <- exec.StreamContext(ctx, options)
	}(ctx, e)
	go func() {
		time.Sleep(time.Second * 1)
		cancel()
	}()
	select {
	case err := <-e:
		return err
	case <-time.After(wait.ForeverTestTimeout):
		return errors.New("execute timeout")
	}
}
```
What I got from the test:

```
test: expected [], got [execute timeout]
```

ccchieh commented:
> ```go
> execute timeout
> ```



> Does the really work?`StreamContext`
> 
> I tried a test and the context.CancelFunc did not stop the stream. If there is any wrong in the test code, please point it. Thanks
> 
> ```go
> func attach2ServerContext(rawURL string, options StreamOptions) error {
> 	uri, _ := url.Parse(rawURL)
> 	exec, err := NewSPDYExecutor(&rest.Config{Host: uri.Host}, "POST", uri)
> 	if err != nil {
> 		return err
> 	}
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> 	e := make(chan error)
> 	go func(ctx context.Context, e chan error) {
> 		e <- exec.StreamContext(ctx, options)
> 	}(ctx, e)
> 	go func() {
> 		time.Sleep(time.Second * 1)
> 		cancel()
> 	}()
> 	select {
> 	case err := <-e:
> 		return err
> 	case <-time.After(wait.ForeverTestTimeout):
> 		return errors.New("execute timeout")
> 	}
> }
> ```
> 
> What I got from the test:
> 
> ```
> test: expected [], got [execute timeout]
> ```

May need to use ctx.Done() :)

```go

	select {
	case <-ctx.Done():
		fmt.Println("Done")
		return
	case <-time.After(wait.ForeverTestTimeout):
		return
	}
```

pangsq commented:
@ccchieh emm, I think the `ctx.Done()` does not mean `exec.StreamContext(ctx, options)` also quit. After the `ctx.Done()`, the connection to HTTP server still exists and the `StreamContext` still hangs.

ccchieh commented:
/retest

ccchieh commented:
@pangsq Thank you for pointing out this problem, I fixed it in the code.

SergeyKanzhelev commented:
/remove sig/node

doesn't seem related to sig-node

neolit123 commented:
/remove-area kubeadm
/remove-sig cluster-lifecycle

ccchieh commented:
/remove-area code-generation
/remove-sig apps 

/remove-area kubectl
/remove-sig node

/remove-area kubelet
/remove-sig testing

ccchieh commented:
/remove-area cloudprovider
/remove-area provider/gcp
/remove-area dependency
/remove-area test

/remove-sig auth

lavalamp commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -140,3 +142,50 @@ func (e *streamExecutor) Stream(options StreamOptions) error {
+
+// StreamContext opens a protocol streamer to the server and streams until a client closes
+// the connection or the server disconnects, but it can be closed through the Context.
+func (e *streamExecutor) StreamContext(ctx context.Context, options StreamOptions) error {
```
Can you modify the existing method such that it makes a context and calls this method? That way there won't be so much code duplication and the only tests you'll have to write are those verifying that context cancellation works. Speaking of tests, this change needs some :)

ccchieh replied:
Thanks for your suggestion. Do I need to add Context to the `Stream` method instead of using `StreamContext`? If so, then I will modify the test.

lavalamp commented on staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go:
```diff
@@ -140,3 +142,50 @@ func (e *streamExecutor) Stream(options StreamOptions) error {
+	}
+	conn, protocol, err := spdy.Negotiate(
+		e.upgrader,
+		&http.Client{Transport: e.transport},
```
(I know you copied this but) http.Client caches connections, IIRC? are the connections flushed? It doesn't seem good to construct one of these every time, why is it done that way?

ccchieh replied:
Sorry, I’m still a novice, so I don’t know much.😢

enj commented:
/remove-sig auth
